apps/emqx/src/emqx_router.erl

```diff
@@ -679,31 +681,89 @@ deinit_schema() ->
     _ = persistent_term:erase(?PT_SCHEMA_VSN),
     ok.
 
--spec choose_schema_vsn(schemavsn()) -> schemavsn().
-choose_schema_vsn(ConfType) ->
-    IsEmptyIndex = emqx_trie:empty(),
-    IsEmptyFilters = is_empty(?ROUTE_TAB_FILTERS),
-    case {IsEmptyIndex, IsEmptyFilters} of
-        {true, true} ->
-            ConfType;
-        {false, true} ->
-            v1;
-        {true, false} ->
-            v2;
-        {false, false} ->
+-spec discover_cluster_schema_vsn() -> schemavsn() | undefined.
+discover_cluster_schema_vsn() ->
+    discover_cluster_schema_vsn(emqx:running_nodes() -- [node()]).
+
+discover_cluster_schema_vsn([]) ->
+    undefined;
+discover_cluster_schema_vsn(Nodes) ->
+    Responses = lists:zipwith(
+        fun
+            (Node, {ok, Schema}) ->
+                {Node, Schema, configured};
+            (Node, {error, {exception, undef, _Stacktrace}}) ->
+                %% No such function on the remote node, assuming it doesn't know about v2 routing.
+                {Node, v1, legacy};
+            (Node, {error, {exception, badarg, _Stacktrace}}) ->
+                %% Likely, persistent term is not defined yet.
+                {Node, unknown, starting};
+            (Node, Error) ->
+                {Node, unknown, Error}
+        end,
+        Nodes,
+        emqx_router_proto_v1:get_routing_schema_vsn(Nodes)
+    ),
+    case lists:usort([Vsn || {_Node, Vsn, _} <- Responses, Vsn /= unknown]) of
+        [Vsn] when Vsn =:= v1; Vsn =:= v2 ->
+            Vsn;
+        [] ->
+            ?SLOG(notice, #{
+                msg => "cluster_routing_schema_discovery_failed",
+                responses => Responses,
+                reason =>
+                    "Could not determine configured routing storage schema in the cluster."
+            }),
+            undefined;
+        [_ | _] ->
+            ?SLOG(critical, #{
+                msg => "conflicting_routing_schemas_configured_in_cluster",
+                responses => Responses,
+                reason =>
+                    "There are nodes in the cluster with different configured routing "
+                    "storage schemas. This probably means that some nodes use v1 schema "
+                    "and some use v2, independently of each other. The routing is likely "
+                    "broken. Manual intervention required."
```

Comment:
maybe add a `force_v2` option and advise to: 
1. ensure software is upgraded to {current_version}
2. change config to add config `node.routing.storage_schema=force_v2`
3. restart the node to join cluster
And repeat steps for all nodes.